src/Compilers/CSharp/Portable/Parser/Lexer.cs

```diff
@@ -595,175 +545,62 @@ private void ScanSyntaxToken(ref TokenInfo info)
-                    {
-                        TextWindow.AdvanceChar();
-
-                        if (TextWindow.PeekChar() == '=')
-                        {
-                            TextWindow.AdvanceChar();
-                            info.Kind = SyntaxKind.QuestionQuestionEqualsToken;
-                        }
-                        else
-                        {
-                            info.Kind = SyntaxKind.QuestionQuestionToken;
-                        }
-                    }
-                    else
-                    {
-                        info.Kind = SyntaxKind.QuestionToken;
-                    }
-
+                    info.Kind = TextWindow.TryAdvance('?')
+                        ? TextWindow.TryAdvance('=') ? SyntaxKind.QuestionQuestionEqualsToken : SyntaxKind.QuestionQuestionToken
+                        : SyntaxKind.QuestionToken;
                     break;
 
                 case '+':
                     TextWindow.AdvanceChar();
-                    if ((character = TextWindow.PeekChar()) == '=')
-                    {
-                        TextWindow.AdvanceChar();
-                        info.Kind = SyntaxKind.PlusEqualsToken;
-                    }
-                    else if (character == '+')
-                    {
-                        TextWindow.AdvanceChar();
-                        info.Kind = SyntaxKind.PlusPlusToken;
-                    }
-                    else
-                    {
-                        info.Kind = SyntaxKind.PlusToken;
-                    }
-
+                    info.Kind =
+                        TextWindow.TryAdvance('=') ? SyntaxKind.PlusEqualsToken :
+                        TextWindow.TryAdvance('+') ? SyntaxKind.PlusPlusToken : SyntaxKind.PlusToken;
                     break;
 
                 case '-':
                     TextWindow.AdvanceChar();
-                    if ((character = TextWindow.PeekChar()) == '=')
-                    {
-                        TextWindow.AdvanceChar();
-                        info.Kind = SyntaxKind.MinusEqualsToken;
-                    }
-                    else if (character == '-')
-                    {
-                        TextWindow.AdvanceChar();
-                        info.Kind = SyntaxKind.MinusMinusToken;
-                    }
-                    else if (character == '>')
-                    {
-                        TextWindow.AdvanceChar();
-                        info.Kind = SyntaxKind.MinusGreaterThanToken;
-                    }
-                    else
-                    {
-                        info.Kind = SyntaxKind.MinusToken;
-                    }
-
+                    info.Kind =
+                        TextWindow.TryAdvance('=') ? SyntaxKind.MinusEqualsToken :
+                        TextWindow.TryAdvance('-') ? SyntaxKind.MinusMinusToken :
+                        TextWindow.TryAdvance('>') ? SyntaxKind.MinusGreaterThanToken : SyntaxKind.MinusToken;
```

Comment:
for example @333fred i find this very easy to read/understand, with the TryAdvances all lined up in this fashion :)

Reply:
I've seen more and more of this, maybe because it's easier to format.
```cs
                    info.Kind = 0 switch
                    {
                        _ when TextWindow.TryAdvance('=') => SyntaxKind.MinusEqualsToken,
                        _ when TextWindow.TryAdvance('-') => SyntaxKind.MinusMinusToken,
                        _ when TextWindow.TryAdvance('>') => SyntaxKind.MinusGreaterThanToken,
                        _ => SyntaxKind.MinusToken
                    };
```

Reply:
yeah, i hate taht.  it def feels liek an abuse given that both the value being switched on, and the actual pattern are meaningless.  I'd rather we had a nice construct for this versus that :)

Reply:
not if merged with the containing switch, but I don't know how you would feel about sideeffecting when clauses.
```cs   
                        case '-' when TextWindow.TryAdvance('='):
                            node.Kind = SyntaxKind.MinusEqualsToken;
                            break;
                        case '-' when TextWindow.TryAdvance('-'):
                            node.Kind = SyntaxKind.MinusMinusToken;
                            break;
                        case '-' when TextWindow.TryAdvance('>'):
                            node.Kind = SyntaxKind.MinusGreaterThanToken;
                            break;
                        case '-'
                            node.Kind = SyntaxKind.MinusToken;
                            break;
```

Reply:
i'm not personally a fan (since the 'when' mutates state).  But i'll defer to @333fred for his thoughts there.  I personally prefer my patterns (And when clauses) to be stateless.  and the mutation then happens in the body.  But i'm def more amenable to that form than the original you posted :)

Reply:
`when` should absolutely not mutate state.